crates/net/network/src/session/active.rs

```diff
@@ -810,6 +878,46 @@ mod tests {
         rx.await.unwrap();
     }
 
+    #[tokio::test(flavor = "multi_thread")]
+    async fn test_request_timeout() {
+        reth_tracing::init_test_tracing();
+
+        let mut builder = SessionBuilder::default();
+
+        let listener = TcpListener::bind("127.0.0.1:0").await.unwrap();
+        let local_addr = listener.local_addr().unwrap();
+
+        let request_timeout = Duration::from_millis(500);
+        let drop_timeout = Duration::from_millis(1500);
+
+        let fut = builder.with_client_stream(local_addr, move |client_stream| async move {
+            let _client_stream = client_stream;
+            tokio::time::sleep(drop_timeout * 100).await;
+        });
+        tokio::task::spawn(fut);
+
+        let (incoming, _) = listener.accept().await.unwrap();
+        let mut session = builder.connect_incoming(incoming).await;
+        session
+            .internal_request_timeout
+            .store(request_timeout.as_millis() as u64, Ordering::Relaxed);
+        session.protocol_breach_request_timeout = drop_timeout;
+        let (tx, rx) = oneshot::channel();
+        let req = PeerRequest::GetBlockBodies { request: GetBlockBodies(vec![]), response: tx };
+        session.on_peer_request(req, Instant::now());
+        tokio::spawn(session);
+
+        let err = rx.await.unwrap().unwrap_err();
+        assert_eq!(err, RequestError::Timeout);
+
+        // wait for protocol breach error
+        let msg = builder.active_session_rx.next().await.unwrap();
+        match msg {
+            ActiveSessionMessage::ProtocolBreach { .. } => {}
+            _ => unreachable!(),
+        }
```

Comment:
> we still accept the response?

we keep waiting for a response until `protocol_breach_request_timeout`, if no response arrived then this is considered a protocol violation, see also:

https://github.com/ethereum/go-ethereum/blob/bd6a05e1eea340cead56478f7fc81cade80a3cad/eth/downloader/fetchers_concurrent.go#L30-L34

but we still answer internal request with a timeout error early so it can be re-requested (syncing)